core/src/main/java/com/graphhopper/routing/util/HeadingEdgeFilter.java

```diff
@@ -18,6 +18,13 @@ public HeadingEdgeFilter(EdgeFilter directedEdgeFilter, double heading, GHPoint
 
     @Override
     public boolean accept(EdgeIteratorState edgeState) {
-        return HeadingResolver.isHeadingNearlyParallel(edgeState, directedEdgeFilter, heading, pointNearHeading);
+        final double tolerance = 30;
+        double headingOfEdge = HeadingResolver.getHeadingOfGeometryNearPoint(edgeState, pointNearHeading);
+        if (Double.isNaN(headingOfEdge))
+            // this edge is too far away. we do not accept it.
+            return false;
```

Comment:
If the found edge is too far away from the query point then we should not snap to it only because of the correct heading. See discussion regarding gpx_accuracy in the original issue.